### PR TITLE
Renderer PR

### PR DESCRIPTION
**Pull request by llGuy**

### CODE REVIEW

File: include/madrona/importer.hpp [hunk 1]
```diff
@@ -48,25 +84,68 @@ struct SourceInstance {
...
 
     static Optional<ImportedAssets> importFromDisk(
         Span<const char * const> asset_paths,
         Span<char> err_buf = { nullptr, 0 },
-        bool one_object_per_asset = false);
+        bool one_object_per_asset = false,
+        bool generate_mesh_bvhs = false);
+
+    // Unfinished but provides just enough to support BVH raytracing.
+    static Optional<GPUGeometryData> makeGPUData(
```
Comment (shacklettbp) [inline include/madrona/importer.hpp:147]: Again shouldn't be here.

File: include/madrona/math.inl [hunk 2]
```diff
@@ -1503,6 +1503,29 @@ float AABB::distance2(const AABB &o) const
...
+inline Vector3 AABB::centroid() const
+{
+    return 0.5f * (pMin + pMax);
+}
+
+inline int AABB::maxDimension() const
+{
+    Vector3 d = pMax - pMin;
+
+    if (d.x > d.y && d.x > d.z) return 0;
```
Comment (shacklettbp) [inline include/madrona/math.inl:1515]: {} please

File: include/madrona/math.inl [hunk 3]
```diff
@@ -1503,6 +1503,29 @@ float AABB::distance2(const AABB &o) const
...
+    else if (d.y > d.z)         return 1;
+    else                        return 2;
+}
+
+inline Vector3 AABB::offset(const Vector3 &p) const
+{
+    Vector3 o = p - pMin;
+    if (pMax.x > pMin.x) o.x /= pMax.x - pMin.x;
```
Comment (shacklettbp) [inline include/madrona/math.inl:1523]: {}

File: include/madrona/mesh_bvh.hpp [hunk 4]
```diff
@@ -1,30 +1,66 @@
 #pragma once
 
-#include <madrona/types.hpp>
-#include <madrona/geo.hpp>
+#include "types.hpp"
```
Comment (shacklettbp) [inline include/madrona/mesh_bvh.hpp:3]: Why are these changed?

File: include/madrona/mw_gpu.hpp [hunk 5]
```diff
@@ -45,6 +45,16 @@ struct StateConfig {
 
     // Number of exported ECS components
     uint32_t numExportedBuffers;
+
+    // This is required if the BVH is to be invoked (holds object information).
+    imp::ImportedAssets::GPUGeometryData *geometryData;
+
+    Span<const imp::SourceMaterial> materials;
```
Comment (shacklettbp) [inline include/madrona/mw_gpu.hpp:52]: I'm not a big fan of this, the import namespace is intended to represent source assets that then get preprocessed into some final runtime form. This might not need to be changed for merge though.

File: include/madrona/mw_gpu.hpp [hunk 6]
```diff
@@ -45,6 +45,16 @@ struct StateConfig {
...
+    Span<const imp::SourceTexture> textures;
+
+    uint32_t raycastOutputResolution;
+
+    float nearSphere;
```
Comment (shacklettbp) [inline include/madrona/mw_gpu.hpp:57]: What is this?

File: include/madrona/physics_assets.hpp [hunk 7]
```diff
@@ -67,7 +67,7 @@ struct MeshBVHBuilder {
     static void * build(
         Span<const imp::SourceMesh> src_meshes,
         StackAlloc &tmp_alloc,
-        MeshBVH *out_bvh,
+        render::MeshBVH *out_bvh,
```
Comment (shacklettbp) [inline include/madrona/physics_assets.hpp:70]: Why has MeshBVH been moved to the render namespace? MeshBVH is used outside the renderer for envs that need to deal with collisions / geo queries on triangle meshes.

File: include/madrona/render/ecs.hpp [hunk 8]
```diff
@@ -21,6 +21,77 @@ struct Renderable {
     Entity renderEntity;
 };
 
+
+
+
+// TODO: Make sure to move this to private headers which can be
```
Comment (shacklettbp) [inline include/madrona/render/ecs.hpp:27]: Either fix this todo or move this all under an impl namespace or something

File: src/importer/gltf.cpp [hunk 9]
```diff
@@ -1323,6 +1327,11 @@ static bool gltfParseMesh(
     const GLTFMesh &gltf_mesh = loader.meshes[mesh_idx];
 
     DynArray<SourceMesh> meshes(1);
+
+    if (gltf_mesh.numPrims > 1) {
```
Comment (shacklettbp) [inline src/importer/gltf.cpp:1331]: WTF?

File: src/importer/gltf.cpp [hunk 10]
```diff
@@ -1630,22 +1679,29 @@ static bool gltfImportAssets(LoaderData &loader,
     DynArray<SourceMesh> merged_meshes(total_new_meshes);
     DynArray<math::Vector3> new_positions_arr(total_new_vertices);
     DynArray<math::Vector3> new_normals_arr(total_new_normals);
+    DynArray<math::Vector2> new_uvs_arr(total_new_uvs);
+    DynArray<uint32_t> new_mats_arr(total_new_mats);
```
Comment (shacklettbp) [inline src/importer/gltf.cpp:1683]: I don't like that all this mat ID stuff has made it all the way down into the importer code. Can this be computed *after* import while setting up the MeshBVH for rendering? Separation of concerns, etc?

File: src/importer/importer.cpp [hunk 11]
```diff
@@ -13,23 +16,123 @@
...
+
+#ifdef MADRONA_CUDA_SUPPORT
+#include <madrona/cuda_utils.hpp>
+#endif
+
 namespace madrona::imp {
 
 using namespace math;
 
+bool loadCache(const char* location, DynArray<render::MeshBVH>& bvhs_out){
```
Comment (shacklettbp) [inline src/importer/importer.cpp:29]: Again let's pull ALL of this MeshBVH / bvh build stuff out of importer and out of the importer lib.

File: src/mw/cuda_exec.cpp [hunk 12]
```diff
@@ -23,6 +23,24 @@
 
 #include "cpp_compile.hpp"
 
+#include <stb_image.h>
+#define KHRONOS_STATIC
+#include <ktx.h>
```
Comment (shacklettbp) [inline src/mw/cuda_exec.cpp:28]: Why are we including KTX in cuda_exec.cpp?

File: src/mw/cuda_exec.cpp [hunk 13]
```diff
@@ -1270,6 +1606,263 @@ static void gpuVMAllocatorThread(HostChannel *channel, CUcontext cu_ctx)
     }
 }
 
+struct MaterialData {
```
Comment (shacklettbp) [inline src/mw/cuda_exec.cpp:1609]: This should not be happening in cuda exec

File: src/mw/cuda_exec.cpp [hunk 14]
```diff
@@ -1458,6 +2068,53 @@ static GPUEngineState initEngineAndUserState(
...
 
+    { // Setup the BVH parameters in the __constant__ block of the bvh module
+        // Pass this to the ECS to fill in
+        MaterialData mat_data = initMaterialData(
+                materials, num_materials,
+                textures, num_textures);
+
+        auto params_tmp = cu::allocGPU(sizeof(mwGPU::madrona::BVHParams));
+
+        printf("From CPU, bvh params temporary is in %p (data %p)\n",
```
Comment (shacklettbp) [inline src/mw/cuda_exec.cpp:2079]: No prints during load.

File: src/mw/device/bvh_print.cpp [hunk 15]
```diff
@@ -0,0 +1,86 @@
+#pragma once
```
Comment (shacklettbp) [inline src/mw/device/bvh_print.cpp:1]: Is this just a carbon copy of the other host_print implementation? Why?

File: src/mw/device/memory.cpp [hunk 16]
```diff
@@ -1,8 +1,22 @@
...
 #include <madrona/utils.hpp>
 
+#include <madrona/mw_gpu/host_print.hpp>
+
+#if defined (MADRONA_MWGPU_BVH_MODULE)
+#define MADRONA_MWGPU_MAX_BLOCKS_PER_SM 4
+#include <madrona/bvh.hpp>
+#endif
+
+#define LOG(...) mwGPU::HostPrint::log(__VA_ARGS__)
```
Comment (shacklettbp) [inline src/mw/device/memory.cpp:12]: This needs to go. MADRONA_MWGPU_LOG bare minimum.

File: src/mw/device/sort_archetype.cpp [hunk 17]
```diff
@@ -982,6 +982,29 @@ void SortArchetypeNodeBase::sortSetup(int32_t)
...
+        state_mgr->archetypeClearNeedsSort(archetypeID);
+    } else {
+        // If this isn't a world sort, this sort will scramble the entities
+        // across worlds and therefore we need to set the needs sort
+        // (which denotes whether the entity needs to be sorted by world)
+        // to true
+        state_mgr->archetypeSetNeedsSort(archetypeID);
+    }
+
+#if 0
```
Comment (shacklettbp) [inline src/mw/device/sort_archetype.cpp:1007]: The sort is so critical to system correctness that let's not leave random commented stuff out in it. If the above code correctly replaces the commented out code, delete the commented out code.

File: src/render/ecs_system.cpp [hunk 18]
```diff
@@ -35,8 +37,45 @@ struct RenderingSystemState {
...
+    x = (x | (x << 16)) & 0b00000011000000000000000011111111;
+    x = (x | (x << 8)) & 0b00000011000000001111000000001111;
+    x = (x | (x << 4)) & 0b00000011000011000011000011000011;
+    x = (x | (x << 2)) & 0b00001001001001001001001001001001;
+
+    return x;
+}
+
+uint32_t encodeMorton3(const Vector3 &v) {
+    return (leftShift3(*((uint32_t *)&v.z)) << 2) |
```
Comment (shacklettbp) [inline src/render/ecs_system.cpp:62]: This is illegal. use std::bit_cast. But I think this is not actually the correct way to do morton codes. Usually I see the FP values converted to some fixed point spacing and then merged together. Unless the leftShift3 is handling this correctly above? You're going to be mixing together exponents and mantissa.

File: src/render/render_common.hpp [hunk 19]
```diff
@@ -24,9 +24,9 @@ namespace InternalConfig {
 inline constexpr uint32_t numFrames = 2;
 inline constexpr uint32_t initMaxTransforms = 100000;
 inline constexpr uint32_t initMaxMatIndices = 100000;
-inline constexpr uint32_t shadowMapSize = 4096;
+inline constexpr uint32_t shadowMapSize = 128;
```
Comment (shacklettbp) [inline src/render/render_common.hpp:27]: Can we bump this back up so shadows don't look horrible in the viewer?

File: src/render/shader/compiler.cpp [hunk 20]
```diff
@@ -265,7 +265,7 @@ static HeapArray<uint32_t> hlslToSPV(
     DynArray<LPCWSTR> dxc_args {
         lshader_path.data(),
         DXC_ARG_WARNINGS_ARE_ERRORS,
-        DXC_ARG_DEBUG,
```
Comment (shacklettbp) [inline src/render/shader/compiler.cpp:268]: Why'd this change?

File: include/madrona/mw_gpu.hpp (hunk 5 above)
Comment (llGuy) [reply include/madrona/mw_gpu.hpp:52]: this is a mistake - not supposed to be there.

File: include/madrona/mw_gpu.hpp (hunk 6 above)
Comment (llGuy) [reply include/madrona/mw_gpu.hpp:57]: again, not supposed to be there - this config is now moved to RenderConfig (it's now nearPlane - I called it near sphere because it's technically a sphere with the tracing).

File: include/madrona/render/ecs.hpp (hunk 8 above)
Comment (llGuy) [reply include/madrona/render/ecs.hpp:27]: keeping it public for now - can be useful sometimes

File: src/importer/gltf.cpp (hunk 9 above)
Comment (llGuy) [reply src/importer/gltf.cpp:1331]: debugging.

File: src/importer/gltf.cpp (hunk 10 above)
Comment (llGuy) [reply src/importer/gltf.cpp:1672]: this is now nuked - no need for it anymore. my bad a lot of this is remnants of some of the experiments we did where we merged the entire scene into one mesh i think. anyway, it's gone now

File: src/mw/cuda_exec.cpp (hunk 12 above)
Comment (llGuy) [reply src/mw/cuda_exec.cpp]: this is resolved. now we have a new scheme that has been pushed for handling ktx

File: src/mw/device/bvh_print.cpp (hunk 15 above)
Comment (llGuy) [reply src/mw/device/bvh_print.cpp:1]: because of the MADRONA_MWGPU_MAX_BLOCKS_PER_SM #define I forgot the exact detail why but it was required for that reason.

Comment (llGuy) [reply src/mw/device/bvh_print.cpp:1]: can fix that with something I do for mw/device/memory.cpp

Comment (llGuy) [reply src/mw/device/bvh_print.cpp:1]: fixed now

File: src/mw/device/sort_archetype.cpp (hunk 17 above)
Comment (llGuy) [reply src/mw/device/sort_archetype.cpp:1007]: resolved

File: src/render/ecs_system.cpp (hunk 18 above)
Comment (llGuy) [reply src/render/ecs_system.cpp:62]: thought so too but https://stackoverflow.com/questions/26856268/morton-index-from-2d-point-with-floats

File: src/render/render_common.hpp (hunk 19 above)
Comment (llGuy) [reply src/render/render_common.hpp:27]: yes

File: src/render/shader/compiler.cpp (hunk 20 above)
Comment (llGuy) [reply src/render/shader/compiler.cpp:268]: dont remmeber but can undo

File: include/madrona/importer.hpp [hunk 21]
```diff
@@ -25,8 +29,59 @@ struct SourceObject {
...
 struct SourceTexture {
-    const char *path;
+    TextureLoadInfo info;
+
+    union {
+        const char *path;
+        PixelBufferInfo pix_info;
+    };
+
+    SourceTexture()
```
Comment (shacklettbp) [inline include/madrona/importer.hpp:66]: Can we move all these into an inl file?

File: external/CMakeLists.txt [hunk 22]
```diff
@@ -117,6 +117,22 @@ target_include_directories(stb SYSTEM INTERFACE
     ${CMAKE_CURRENT_SOURCE_DIR}/stb
 )
 
+#file(GLOB_RECURSE KTX_SOURCES ${CMAKE_CURRENT_SOURCE_DIR}/KTX-Software/include/*.h
```
Comment (shacklettbp) [inline external/CMakeLists.txt:120]: Delete

File: include/madrona/importer.hpp [hunk 23]
```diff
@@ -48,25 +103,60 @@ struct SourceInstance {
...
+    EmbreeLoader(EmbreeLoader &&) = default;
+    ~EmbreeLoader();
+
+    std::unique_ptr<Impl> impl_;
+
+    Optional<MeshBVH> load(const SourceObject &obj,
+            const DynArray<SourceMaterial> &materials);
+};
+
+struct SourceAssetInfo {
```
Comment (shacklettbp) [inline include/madrona/importer.hpp:119]: Is this specifically for textures? Why is a level of abstraction *above* an object necessary? An Object is already composed of multiple meshes.

File: include/madrona/mw_gpu.hpp [hunk 24]
```diff
@@ -69,6 +69,29 @@ struct CompileConfig {
     OptMode optMode = OptMode::LTO;
 };
 
+struct RenderConfig {
+    // Until we support custom rendering functions, for now we must toggle
+    // between rendering color or depth.
+    enum class RenderMode : uint32_t {
+        Color,
+        Depth,
+        None
```
Comment (shacklettbp) [inline include/madrona/mw_gpu.hpp:78]: Trailing comma please

File: include/madrona/mw_gpu.hpp [hunk 25]
```diff
@@ -93,8 +116,9 @@ class MWCudaExecutor {
     static CUcontext initCUDA(int gpu_id);
 
     MWCudaExecutor(const StateConfig &state_cfg,
-                                        const CompileConfig &compile_cfg,
-                                        CUcontext cu_ctx);
+                   const CompileConfig &compile_cfg,
+                   const RenderConfig &render_cfg,
```
Comment (shacklettbp) [inline include/madrona/mw_gpu.hpp:120]: Let's move RenderConfig to the last argument and have it be a default argument that disables the renderer. Trying to minimize churn for existing users.

File: include/madrona/render/render_mgr.hpp [hunk 26]
```diff
@@ -17,8 +17,15 @@ struct RenderContext;
 class RenderManager {
 public:
     struct Config {
+        enum RenderMode {
+            Color,
+            Depth
```
Comment (shacklettbp) [inline include/madrona/render/render_mgr.hpp:22]: Trailing comma

File: src/importer/embree.hpp [hunk 27]
```diff
@@ -0,0 +1,9 @@
+#pragma once
```
Comment (shacklettbp) [inline src/importer/embree.hpp:1]: What's this file?

File: src/importer/gltf.cpp [hunk 28]
```diff
@@ -1518,7 +1545,7 @@ static bool gltfParseMesh(
             .faceMaterials = nullptr,
             .numVertices = num_vertices,
             .numFaces = num_faces,
-            .materialIDX = 0, // FIXME
+            .materialIDX = prim.materialIdx, // FIXME
```
Comment (shacklettbp) [inline src/importer/gltf.cpp:1548]: Does this still need a fixme?

File: src/importer/gltf.cpp [hunk 29]
```diff
@@ -1622,6 +1652,10 @@ static bool gltfImportAssets(LoaderData &loader,
             if (src_mesh.tangentAndSigns) {
                 total_new_tangents += src_mesh.numVertices;
             }
+
+            if(src_mesh.uvs){
```
Comment (shacklettbp) [inline src/importer/gltf.cpp:1656]: spaces

File: src/importer/gltf.cpp [hunk 30]
```diff
@@ -1672,6 +1713,12 @@ static bool gltfImportAssets(LoaderData &loader,
                         src_mesh.tangentAndSigns[i].w,
                     });
                 }
+
+                if(src_mesh.uvs){
```
Comment (shacklettbp) [inline src/importer/gltf.cpp:1717]: spaces

File: src/importer/gltf.cpp [hunk 31]
```diff
@@ -1691,11 +1738,17 @@ static bool gltfImportAssets(LoaderData &loader,
         }
     }
 
+    for(SourceMesh& mesh:merged_meshes){
```
Comment (shacklettbp) [inline src/importer/gltf.cpp:1741]: spaces

File: src/importer/gltf.cpp [hunk 32]
```diff
@@ -1716,6 +1769,56 @@ static bool gltfImportAssets(LoaderData &loader,
     imported.geoData.meshArrays.emplace_back(std::move(merged_meshes));
     imported.geoData.positionArrays.emplace_back(std::move(new_positions_arr));
     imported.geoData.normalArrays.emplace_back(std::move(new_normals_arr));
+    imported.geoData.uvArrays.emplace_back(std::move(new_uvs_arr));
+
+    CountT prev_img_idx = imported.imgData.imageArrays.size();
+    CountT prev_tex_idx = imported.texture.size();
+
+    for(const auto& image : loader.images){
```
Comment (shacklettbp) [inline src/importer/gltf.cpp:1777]: spaces

File: src/importer/gltf.cpp [hunk 33]
```diff
@@ -1716,6 +1769,56 @@ static bool gltfImportAssets(LoaderData &loader,
...
+        GLTFBufferView view = loader.bufferViews[image.viewIdx];
+        auto *image_memory = (uint8_t*)malloc(view.numBytes);
+        memcpy(image_memory, loader.buffers[view.bufferIdx].dataPtr + view.offset, view.numBytes);
+
+        TextureFormat format;
+        switch(image.type){
```
Comment (shacklettbp) [inline src/importer/gltf.cpp:1783]: spaces

File: src/importer/gltf.cpp [hunk 34]
```diff
@@ -1716,6 +1769,56 @@ static bool gltfImportAssets(LoaderData &loader,
...
+            default:
+                format = TextureFormat::JPG;
+                break;
+        }
+
+        imported.imgData.imageArrays.emplace_back(BackingImageData {
+            .imageData = image_memory, .imageSize = view.numBytes,.format = format});
+    }
+
+    for(const auto& texture : loader.textures){
```
Comment (shacklettbp) [inline src/importer/gltf.cpp:1802]: spaces

File: src/importer/gltf.cpp [hunk 35]
```diff
@@ -1716,6 +1769,56 @@ static bool gltfImportAssets(LoaderData &loader,
...
+        uint32_t backingIndex = (texture.sourceIdx + prev_img_idx);
+
+        imported.texture.emplace_back(SourceTexture(
+                PixelBufferInfo{.backingDataIndex = backingIndex}));
+    }
+
+    for(const auto& material : loader.materials){
```
Comment (shacklettbp) [inline src/importer/gltf.cpp:1809]: spaces

File: src/importer/gltf.cpp [hunk 36]
```diff
@@ -1716,6 +1769,56 @@ static bool gltfImportAssets(LoaderData &loader,
...
+        int32_t texture_id = material.baseColorIdx;
+        if(texture_id != -1){
```
Comment (shacklettbp) [inline src/importer/gltf.cpp:1811]: spaces

File: src/importer/gltf.cpp [hunk 37]
```diff
@@ -1716,6 +1769,56 @@ static bool gltfImportAssets(LoaderData &loader,
...
+            texture_id += prev_tex_idx;
+        }
+        SourceMaterial s_mat = {
+            .color = material.baseColor,
+            //.color = math::Vector4{1.f, 1.f, 1.f, 1.f},
+            .textureIdx = texture_id,
+            .roughness=material.roughness,
+            .metalness=material.metallic};
```
Comment (shacklettbp) [inline src/importer/gltf.cpp:1819]: Add trailing comma, } on newline please

File: src/importer/importer.cpp [hunk 38]
```diff
@@ -16,10 +19,122 @@
...
+        return std::hash<std::string_view>{}({reinterpret_cast<const char*>(x.data()), x.size()});
+    }
+};
+
 namespace madrona::imp {
 
 using namespace math;
 
+void ImportedAssets::postProcessTextures(const char *texture_cache, TextureProcessFunc process_tex_func) {
+    for(SourceTexture& tx: texture) {
```
Comment (shacklettbp) [inline src/importer/importer.cpp:47]: spaces

File: src/importer/importer.cpp [hunk 39]
```diff
@@ -16,10 +19,122 @@
...
+    {
+        return std::hash<std::string_view>{}({reinterpret_cast<const char*>(x.data()), x.size()});
+    }
+};
+
 namespace madrona::imp {
 
 using namespace math;
 
+void ImportedAssets::postProcessTextures(const char *texture_cache, TextureProcessFunc process_tex_func) {
```
Comment (shacklettbp) [inline src/importer/importer.cpp:46]: { on newline

File: src/mw/CMakeLists.txt [hunk 40]
```diff
@@ -29,7 +29,10 @@ target_link_libraries(madrona_mw_gpu
         madrona_mw_core madrona_json
         CUDA::nvrtc CUDA::cuda_driver
         ${CUDA_NVJITLINK_LIBRARY}
+        stb
 )
+get_target_property(OUT madrona_mw_gpu LINK_LIBRARIES)
```
Comment (shacklettbp) [inline src/mw/CMakeLists.txt:34]: What's this?

File: src/mw/CMakeLists.txt [hunk 41]
```diff
@@ -29,7 +29,10 @@ target_link_libraries(madrona_mw_gpu
         madrona_mw_core madrona_json
         CUDA::nvrtc CUDA::cuda_driver
         ${CUDA_NVJITLINK_LIBRARY}
+        stb
```
Comment (shacklettbp) [inline src/mw/CMakeLists.txt:32]: Does this still need to be linked?

File: src/mw/device/sort_archetype.cpp (hunk 17 above)
Comment (shacklettbp) [reply src/mw/device/sort_archetype.cpp:1010]: This doesn't seem resolved? There's still commented out code

File: src/render/render_common.hpp (hunk 19 above)
Comment (shacklettbp) [reply src/render/render_common.hpp:27]: Again, doesn't seem resolved? I still see 128

**Review by shacklettbp — CHANGES_REQUESTED** [review]

File: src/mw/device/sort_archetype.cpp (hunk 17 above)
Comment (llGuy) [reply src/mw/device/sort_archetype.cpp:1010]: sorry now it's removed

File: include/madrona/importer.hpp (hunk 23 above)
Comment (llGuy) [reply include/madrona/importer.hpp:119]: discussed in slack

File: src/importer/embree.hpp (hunk 27 above)
Comment (llGuy) [reply src/importer/embree.hpp:1]: removed

File: src/importer/gltf.cpp (hunk 28 above)
Comment (llGuy) [reply src/importer/gltf.cpp]: no
